# Users' Guide
LinkedSV is a structural variant caller for 10X Genomics (linked-read) sequencing data. It detects deletions, duplications, inversions and translocations using evidence from the barcoded reads. 

## Installation

### Prerequisites
GCC (version >= 4.8).

Python packages: sklearn, scipy, numpy, gzip, psutil, subprocess, bisect, math, argparse.

SAMtools (https://github.com/samtools/samtools ) and BEDTools (https://bedtools.readthedocs.io/en/latest/). 


```
git clone https://github.com/WGLab/LinkedSV.git 
cd LinkedSV/
sh build.sh 
```

## Usage

```
usage: linkedsv.py [-h] -i input.phased_possorted_bam.bam -d output_directory
                   -r ref.fa [-v version] [--gap_region_bed BED]
                   [--black_region_bed BED] [-t num_thread] [-q min_map_qual]
                   [--min_fragment_length INT] [--min_supp_barcodes INT]
                   [--samtools path/to/samtools] [--bedtools path/to/bedtools]
                   [--wgs] [--targeted] [--target_region BED]
                   [--gap_distance_cut_off INT] [--rm_temp_files INT]

```

### Examples: 
`python linkedsv.py -i phased_possorted_bam.bam -d path/to/output_dir/ -r ref.fasta -v hg38 -t 4`

The `phased_possorted_bam.bam` is the full path to the phased_possorted_bam.bam file generated by Longranger.

The `ref.fasta` file is the FASTA file of the reference genome. It should be the same fasta file that was used for alignment.

`ref_version` is used to tell LinkedSV which black_list file and gap_region file should be used. Currently we have generated blacklists for hg19 (style: "chr1"), b37 (style: "1") and hg38 (style: "chr1"). If `ref_version` is not specifed or you are using a different reference file, please generate these files by yourself and specify the `--gap_region_bed` and `--black_region_bed` parameters. 

`-t 4` tells LinkedSV to use 4 threads. We recommend using at least 4 threads to speed up the run. Each thread need 4GB memory.

### Output files of LinkedSV
The output file is the prefix.filtered_svcalls.bedpe file. 

The output file is a TAB-delimited text file. The meaning of each column is described in the following table. 

|Column|Type|Description|
|--:|:----:|:-----------------------------------------|
|1|string|chrom of breakpoint 1|
|2|int|start position of breakpoint 1|
|3|int|end position of breakpoint 1|
|4|string|chrom of breakpoint 2|
|5|int|start position of breakpoint 2|
|6|int|end position of breakpoint 2|
|7|string|SV type inferred from breakpoints|
|8|int|SV length|
|9|int|number of barcodes that support the SV|
|10|int|number of confident short reads that support the SV|
|11|string|type of breakpoint 1|
|12|string|type of breakpoint 2|
|13|float|score of the SV|
|14|float|intermediate score (will not be output in the future versions)|
|15|string|barcodes that support the SV|
