# Users' Guide
LinkedSV is a structural variant caller for 10X Genomics (linked-read) sequencing data. It detects deletions, duplications, inversions and translocations using evidence from the barcoded reads. 

## Installation

### Prerequisites
GCC (version >= 4.4).

Python (version: 2.7)

Python packages: sklearn, scipy, numpy, gzip, psutil, subprocess, bisect, math, argparse.

SAMtools (version >= 1.3, https://github.com/samtools/samtools ) and BEDTools (https://bedtools.readthedocs.io/en/latest/). 

```
git clone https://github.com/WGLab/LinkedSV.git 
cd LinkedSV/
sh build.sh 
```

## Usage

```
usage: linkedsv.py [-h] -i input.phased_possorted_bam.bam -d output_directory
                   -r ref.fa [-v version] [--gap_region_bed BED]
                   [--black_region_bed BED] [-t num_thread] [-q min_map_qual]
                   [--min_fragment_length INT] [--min_supp_barcodes INT]
                   [--samtools path/to/samtools] [--bedtools path/to/bedtools]
                   [--wgs] [--targeted] [--target_region BED]
                   [--gap_distance_cut_off INT] [--rm_temp_files INT]
```

### Examples: 
1. Detection of SVs from whole-genome sequencing. 

```
python linkedsv.py -i phased_possorted_bam.bam -d path/to/output_dir/ -r ref.fasta -v hg38 -t 4
```


The `phased_possorted_bam.bam` is the full path to the phased_possorted_bam.bam file generated by Longranger.

The `ref.fasta` file is the FASTA file of the reference genome. It should be the same fasta file that was used for alignment.

`ref_version` is used to tell LinkedSV which black_list file and gap_region file should be used. Currently we have generated blacklists for hg19 (style: "chr1"), b37 (style: "1") and hg38 (style: "chr1"). If `ref_version` is not specifed or you are using a different reference file, please generate these files by yourself and specify the `--gap_region_bed` and `--black_region_bed` parameters. 

We recommend using at least 4 threads to speed up the run. Each thread need 4GB memory.

2. Detection of SVs from targeted sequencing (e.g. whole-exome sequencing) . 

```
python linkedsv.py -i phased_possorted_bam.bam -d path/to/output_dir/ -r ref.fasta -v hg38 -t 4 --targeted --target_region path/to/target_region.bed
```

`target_region.bed` is a bed file that contains the target regions (capture regions). 


### Output files of LinkedSV

LinkedSV will output two files, `prefix.raw_svcalls.bedpe` and `prefix.filtered_svcalls.bedpe`. In most cases, you only need to look at the `prefix.filtered_svcalls.bedpe`. 

The BEDPE format was defined by bedtools (https://bedtools.readthedocs.io/en/latest/content/general-usage.html). It can be used to concisely describe disjoint genome features, such as structural variations or paired-end sequence alignments.

The `prefix.filtered_svcalls.bedpe` file contains one SV per line with the following tab-delimited columns:

|Column|Description|
:----:|:-----------------------------------------|
|chrom1|chrom of breakpoint 1|
|start1|start position of breakpoint 1|
|stop1|end position of breakpoint 1|
|chrom2|chrom of breakpoint 2|
|start2|start position of breakpoint 2|
|stop2|end position of breakpoint 2|
|sv_type|SV type inferred from breakpoints|
|sv_length|SV length|
|filter|filter. 'PASS' if the call passed all filtering steps.| 
|num_supporting_fragments|number of fragment pairs that support the SV|
|num_supporting_read_pairs|number of read pairs that support the SV|
|endpoint1_type|type of enriched fragment endpoint near breakpoint 1|
|endpoint2_type|type of enriched fragment endpoint near breakpoint 2|
|qual_score|quality score of the SV|
|supporting_barcodes|barcode sequences of the fragments that support the SV|


## Citation
If you use LinkedSV in your work, please cite:
> Li Fang, Charlly Kao, Michael V Gonzalez, Fernanda A Mafra, Renata Pellegrino da Silva, Mingyao Li, Soren Wenzel, Katharina Wimmer, Hakon Hakonarson, Kai Wang. LinkedSV: Detection of mosaic structural variants from linked-read exome and genome sequencing data. bioRxiv 409789; doi: https://doi.org/10.1101/409789

## Getting Help

Please use the [GitHub's Issues page](https://github.com/WGLab/LinkedSV/issues) if you have questions.
